File: ProxyHelper_PreMkVII/api/module.php
<?php namespace pineapple;

class ProxyHelper extends Module
{
    private $tmpFileName = 'iptables_tmp';
    private $statusFile = '/pineapple/modules/ProxyHelper/module.status';

    public function route()
    {
        switch ($this->request->action) {
            case 'aboutInfo':
                $this->aboutInfo();
                break;
            case 'clearRules':
                $this->clearRules();
                break;
            case 'backupRules':
                $this->backupRules();
                break;
            case 'viewBackup':
                $this->viewBackup();
                break;
            case 'restoreBackup':
                $this->restoreBackup();
                break;
            case 'getBackups':
                $this->getBackups();
                break;
            case 'deleteBackup':
                $this->deleteBackup();
                break;
            case 'createProxyRules':
                $this->createProxyRules();
                break;
            case 'setRunningStatus':
                $this->setRunningStatus();
                break;
            case 'getRunningStatus':
                $this->getRunningStatus();
                break;
        }
    }

    protected function aboutInfo()
    {
        $moduleInfo = @json_decode(file_get_contents("/pineapple/modules/ProxyHelper/module.info"));
        $this->response = array('title' => $moduleInfo->title, 'version' => $moduleInfo->version);
    }

    private function clearRules()
    {
        /*
        Clear out the PREROUTING and POSTROUTING chains to remove our rules.
        Just in case this clears out something the user needed, we make and
        restore a backup when toggling the proxy. If this ends up causing issues,
        the code could be updated to parse iptables output and remove only what we added
        */
        exec('iptables -t nat -F PREROUTING', $output, $retval); // Flush all chains
        exec('iptables -t nat -F POSTROUTING', $output, $retval); // Delete every non-builtin chain

        $this->response = array('success' => true,
                                'call' => 'clearRules()');
    }

    private function createProxyRules()
    {
        // Validate the IP address
        if (!filter_var($this->request->dIP, FILTER_VALIDATE_IP)) {
            $this->response = array('success' => false,
                                  'call' => 'createProxyRules()',
                                  'error' => 'The IP address is invalid');
            return;
        }

        // Validate the port number
        if (!filter_var($this->request->dPort, FILTER_VALIDATE_INT)) {
            $this->response = array('success' => false,
                                  'call' => 'createProxyRules()',
                                  'error' => 'The port is invalid (not an integer)');
            return;
        }

        // Validate the port range
        if (intval($this->request->dPort) <= 0 || intval($this->request->dPort) > 65535) {
            $this->response = array('success' => false,
                                  'call' => 'createProxyRules()',
                                  'error' => 'The port is invalid (invalid range)');
            return;
        }

        // Enable IP forwarding
        exec('echo \'1\' > /proc/sys/net/ipv4/ip_forward', $output, $retval);

        // Create the proxy rules
        $destination = $this->request->dIP . ':' . $this->request->dPort;
        exec('iptables -t nat -A PREROUTING -p tcp --dport 80 -j DNAT --to-destination ' . $destination);
        exec('iptables -t nat -A PREROUTING -p tcp --dport 443 -j DNAT --to-destination ' . $destination);
        exec('iptables -t nat -A POSTROUTING -j MASQUERADE');
        $this->response = array('success' => true,
                                'call' => 'createProxyRules()',
                                'IP' => $this->request->dIP,
                                'Port' => $this->request->dPort,
                                'destination' => $destination);
    }

    private function backupRules()
    {
        // create the backup directory if it does not exist
        exec('mkdir -p /pineapple/modules/ProxyHelper/backups/', $output, $retval);
        $filename = '';

        if ($this->request->bAutoBackup) { // Name the auto backup something different to hide from the interface
            $filename = '/pineapple/modules/ProxyHelper/backups/' . $this->tmpFileName;
        } else {
            $filename = '/pineapple/modules/ProxyHelper/backups/iptables_' . date('Y-m-d_H-i-s');
        }

        // backup the current rules
        exec('iptables-save > ' . $filename, $output, $retval);
        $this->response = array('success' => true,
                                'content' => 'iptables-save',
                                'filename' => $filename,
                                'output' => $output,
                                'return' => $retval);
    }

    private function viewBackup()
    {
        exec('cat /pineapple/modules/ProxyHelper/backups/'.$this->request->file, $output);
        $file = $this->request->file;

        if (!empty($output)) {
            $this->response = array('output' => implode("\n", $output), 'file' => $file);
        } else {
            $this->response = array('output' => 'Empty backup...', 'file' => $file);
        }
    }

    private function restoreBackup()
    {

        $filename = '';
        $filename = '/pineapple/modules/ProxyHelper/backups/' . $this->request->file;

        // restore the rules
        exec('iptables-restore < ' . $filename, $output, $retval);
        $this->response = array('success' => true,
                                'content' => 'iptables-restore',
                                'filename' => $filename,
                                'output' => $output,
                                'return' => $retval);
    }

    private function getBackups()
    {
        $baseDir = '/pineapple/modules/ProxyHelper/backups';
        $backups = [];
        $scan = scandir($baseDir);

        foreach ($scan as $file) {
            // Ensure it is a file and not a path
            if (is_file($baseDir . '/' . $file)) {
                // Ignore the auto created iptables backup
                if (basename($file) != $this->tmpFileName) {
                    array_push($backups, $file);
                }
            }
        }

        $this->response = array('backups' => $backups);
    }

    private function deleteBackup()
    {
        exec('rm -rf /pineapple/modules/ProxyHelper/backups/'.$this->request->file);
        $this->response = array('deleteBackup()' => $this->request->file);
    }

    // Create status file that we can check if the user navigates away from the page
    private function setRunningStatus()
    {
        $arr = array('running' => $this->request->running,
                     'proxyIP' => $this->request->proxyIP,
                     'proxyPort' => $this->request->proxyPort);

        $json = json_encode($arr);
        $myfile = fopen($this->statusFile, "w");
        fwrite($myfile, $json);
        fclose($myfile);

        $this->response = array('success' => true,
                                'call' => 'setRunningStatus()',
                                'json' => $json);
    }

    private function getRunningStatus()
    {
        $running = false;
        $proxyIP = '';
        $proxyPort = '';

        if (file_exists($this->statusFile)) {
            $txt = file_get_contents($this->statusFile);
            $json = json_decode($txt);
            $running = $json->running;
            $proxyIP = $json->proxyIP;
            $proxyPort = $json->proxyPort;
        }

        $this->response = array('success' => true,
                                'call' => 'getRunningStatus()',
                                'running' => $running,
                                'proxyIP' => $proxyIP,
                                'proxyPort' => $proxyPort);
    }
}

File: ProxyHelper/projects/ProxyHelper/src/module.py
#!/usr/bin/env python3 
 
import logging 
 
from pineapple.modules import Module, Request 
 
module = Module("ProxyHelper", logging.DEBUG) 

tmpFileName = "iptables_tmp"
statusFile = "/pineapple/modules/ProxyHelper/module.status"
 
@module.handles_action("aboutInfo") 
def aboutInfo(request: Request): 
    f = open("/pineapple/modules/ProxyHelper/module.json")
    data = json.load(f)
    return json.dumps({"title": data["title"], "version": data["version"]})

@module.handles_action("clearRules")
def clearRules(request: Request):
    # Clear out the PREROUTING and POSTROUTING chains to remove our rules.
    # Just in case this clears out something the user needed, we make and
    # restore a backup when toggling the proxy. If this ends up causing issues,
    # the code could be updated to parse iptables output and remove only what we added    
    os.system("iptables -t nat -F PREROUTING")
    os.system("iptables -t nat -F POSTROUTING")
    return json.dumps({"success": True, "call": "clearRules()"})

@module.handles_action("createProxyRules")
def createProxyRules(request: Request):
    if re.search(r'((^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$))', request.dIP) == "":
        return json.dumps({"success": False, "call": "createPoxyRules()", "error": "The IP address is invalid"})
    
    if not isinstance(request.dPort, int):
        return json.dumps({"success": False, "call": "createPoxyRules()", "error": "The port is invalid (not an integer)"})

    if request.dbPort <= 0 or request.dbPort >65535:
        return json.dumps({"success": False, "call": "createPoxyRules()", "error": "The port is invalid (invalid range)"})
    
    os.system("echo '1' > /proc/sys/net/ipv4/ip_forward")
    destination = request.dIP + ":" + request.dPort
    os.system("iptables -t nat -A PREROUTING -p tcp --dport 80 -j DNAT --to-destination " + destination)
    os.system("iptables -t nat -A PREROUTING -p tcp --dport 443 -j DNAT --to-destination " + destination)
    os.system("iptables -t nat -A POSTROUTING -j MASQUERADE")
    return json.dumps({"success": True, "call": "createProxyRules()", "ip": request.dIP, "port": request.dPort, "destination": destination})

@module.handles_action("backupRules")
def backupRules(request: Request):
    os.system("mkdir -p /pineapple/modules/ProxyHelper/backups/")
    filename = ""
    if request.bIsAutoBackup:
        filename = "/pineapple/modules/ProxyHelper/backups/" + tmpFileName
    else:
        filename = '/pineapple/modules/ProxyHelper/backups/iptables_' + datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-S")

    os.system("iptables-save > " + filename)

    return json.dumps({"success": True, "call": "backupRules()", "content": "iptables-save", "filename": filename})

 
if __name__ == "__main__": 
    module.start()